### ISSUE TITLE
Backward gradient not scaled properly

### ISSUE THREAD

**Issue by gouthamvgk** [zzzxxxttt/pytorch_DoReFaNet#4]
In utils/quant_dorefa.py, Line 41
`weight_q = self.uniform_q(x / E) * E` 
The sign values is multiplied with E outside the uniform_quatize function. So when gradient is backpropagated , it will be multiplied by E and then backpropagated to uniform_quantize which will return the gradient unchanged. But the gradient with respect to float weights is samed as gradient with respect to quantized weights in dorefa net. So multiplication by E should happen inside the uniform_quantize function.
The fix is change line 15 to,
```
E = torch.mean(torch.abs(input)).detach()
out = torch.sign(input/E) * E
```
Line 41 to, 
`weight_q = self.uniform_q(x)`

**Comment by zzzxxxttt** [comment]
Thanks for your comment, I think in my implementation the gradient w.r.t float weights are identical to the gradient w.r.t the quantized weights, please correct me if my derivation is wrong.
![dw](https://user-images.githubusercontent.com/39305301/65097714-594fd900-d9f9-11e9-97a6-198023ca5910.png)

**Comment by gouthamvgk** [comment]
Thank you for the clarification, got the point!
